modular/gater/migrate_handler.go

```diff
@@ -300,3 +363,287 @@ func (g *GateModular) getSwapOutApproval(w http.ResponseWriter, r *http.Request)
+	}()
+
+	reqCtx, _ = NewRequestContext(r, g)
+	if allowMigrate, bucketMigrationInfo, _, err = g.verifySignatureAndSP(reqCtx, r.Header.Get(GnfdMigrateBucketMsgHeader)); err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to check migrate bucket auth", "bucket_migration_info", bucketMigrationInfo, "error", err)
+		return
+	}
+	if !allowMigrate {
+		log.CtxErrorw(reqCtx.Context(), "no permission to migrate piece", "bucket_migration_info", bucketMigrationInfo)
+		err = ErrNoPermission
+		return
+	}
+
+	bucketID = bucketMigrationInfo.GetBucketId()
+
+	quota, err := g.baseApp.GfSpClient().GetLatestBucketReadQuota(
+		reqCtx.Context(), bucketID)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to get bucket read quota", "bucket_id",
+			bucketID, "error", err)
+		return
+	}
+
+	bz, err := quota.Marshal()
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to marshal", "bucket_id",
+			bucketID, "error", err)
+		return
+	}
+
+	w.Header().Set(GnfdSignedApprovalMsgHeader, hex.EncodeToString(bz))
+	w.Header().Set(ContentTypeHeader, ContentTypeXMLHeaderValue)
+
+	log.CtxInfow(reqCtx.Context(), "succeed to get latest bucket quota", "bucket_id",
+		bucketID, "quota", quota)
+}
+
+// preMigrateBucketHandler handles the prepare request for bucket migrate
+func (g *GateModular) preMigrateBucketHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		err                 error
+		reqCtx              *RequestContext
+		bucketID            uint64
+		bucketSize          uint64
+		bucketMigrationInfo *gfsptask.GfSpBucketMigrationInfo
+		allowMigrate        bool
+	)
+
+	defer func() {
+		reqCtx.Cancel()
+		if err != nil {
+			reqCtx.SetError(err)
+			reqCtx.SetHTTPCode(int(gfsperrors.MakeGfSpError(err).GetHttpStatusCode()))
+			modelgateway.MakeErrorResponse(w, gfsperrors.MakeGfSpError(err))
+		} else {
+			reqCtx.SetHTTPCode(http.StatusOK)
+		}
+		log.CtxDebugw(reqCtx.Context(), reqCtx.String())
+	}()
+
+	reqCtx, _ = NewRequestContext(r, g)
+	if allowMigrate, bucketMigrationInfo, err = g.checkMigrateBucketQuotaAuth(reqCtx, r.Header.Get(GnfdMigrateBucketMsgHeader)); err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to check migrate bucket auth", "migrate_bucket", bucketMigrationInfo, "error", err)
+		return
+	}
+	if !allowMigrate {
+		log.CtxErrorw(reqCtx.Context(), "no permission to migrate piece", "migrate_bucket", bucketMigrationInfo)
+		err = ErrNoPermission
+		return
+	}
+
+	bucketID = bucketMigrationInfo.GetBucketId()
+
+	// get bucket quota and check, lock quota
+	bucketSize, err = g.getBucketTotalSize(reqCtx.Context(), bucketID)
+	if err != nil {
+		return
+	}
+
+	quota, err := g.baseApp.GfSpClient().GetLatestBucketReadQuota(
+		reqCtx.Context(), bucketID)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to get bucket read quota", "bucket_id",
+			bucketID, "error", err)
+		return
+	}
+
+	// reduce quota, sqldb
+	err = g.baseApp.GfSpClient().DeductQuotaForBucketMigrate(
```

Comment:
If the dest SP retry calling `preMigrateBucket`, would it deduct multiple times?